GitLab config with Redis fails in GitLab CI

##### **Overview of the issue**

JHipster generated project with Redis support fails on GitLab CI, despite Redis container running.

##### **Motivation for or Use Case**

When using Redis, it should be possible to run the CI pipeline on GitLab which supports Redis.

##### **Reproduce the error**

Generate application with Redis support. 

Generate GitLab CI config:

    jhipster ci-cd

Select GitLab, and enable "dockerized"

Deploy the application to GitLab, and run the pipeline. 

Job maven-test fails, because several integration tests fail with:

Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.cache.configuration.Configuration]: Factory method 'jcacheConfiguration' threw exception; nested exception is java.lang.IllegalArgumentException: Redis url should start with redis:// or rediss:// (for SSL connection)
 	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
 	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:651)
 	... 122 common frames omitted
 Caused by: java.lang.IllegalArgumentException: Redis url should start with redis:// or rediss:// (for SSL connection)
 	at org.redisson.misc.RedisURI.<init>(RedisURI.java:40)

Tried to manally set things, but 

##### **Related issues**

None

##### **Suggest a Fix**

Try to see what configuration is missing when running GitLab pipeline, and add that to the gitlab ci-cd code generation for GitLab. 

##### **JHipster Version(s)**

6.8.0

##### **JHipster configuration**

{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.treemarks",
      "nativeLanguage": "nl"
    },
    "jhipsterVersion": "6.8.0",
    "applicationType": "monolith",
    "baseName": "treemarks",
    "packageName": "com.treemarks",
    "packageFolder": "com/treemarks",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "redis",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "embeddableLaunchScript": false,
    "useSass": true,
    "clientPackageManager": "npm",
    "clientFramework": "angularX",
    "clientTheme": "superhero",
    "clientThemeVariant": "primary",
    "creationTimestamp": 1587216333162,
    "testFrameworks": ["cucumber"],
    "jhiPrefix": "jhi",
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "enableTranslation": true,
    "nativeLanguage": "nl",
    "languages": ["nl", "en"],
    "blueprints": []
  }
}

---

If you have a JDL please wrap it in below structure

entity Person {
    givenName String,
    familyName String,
    mboxSha1 String
}

entity Statement {
    subject String,
    predicate String,
    object String
}

relationship OneToMany {
	Person{knows} to Person
}

relationship ManyToOne {
	Statement{owner} to Person
}

---

.gitlab-ci.yml:


```
image: jhipster/jhipster:v6.8.0

#services:
#  - redis

#connect:
#  image: redis
#  script:
#    - redis-cli -h redis PING

cache:
  key: '$CI_COMMIT_REF_NAME'
  paths:
    - .maven/
stages:
  - check
  - build
  - test
  - analyze
  - package
  - release
  - deploy
before_script:
  - export NG_CLI_ANALYTICS="false"
  - export MAVEN_USER_HOME=`pwd`/.maven
  - ./mvnw -ntp com.github.eirslett:frontend-maven-plugin:install-node-and-npm -DnodeVersion=v12.16.1 -DnpmVersion=6.14.2 -Dmaven.repo.local=$MAVEN_USER_HOME
  - ./mvnw -ntp com.github.eirslett:frontend-maven-plugin:npm -Dmaven.repo.local=$MAVEN_USER_HOME

nohttp:
  stage: check
  script:
    - ./mvnw -ntp checkstyle:check -Dmaven.repo.local=$MAVEN_USER_HOME

maven-compile:
  stage: build
  script:
    - ./mvnw -ntp compile -Dmaven.repo.local=$MAVEN_USER_HOME
  artifacts:
    paths:
      - target/classes/
      - target/generated-sources/
    expire_in: 1 day

maven-test:
  stage: test
  script:
    - ./mvnw -ntp verify -Dmaven.repo.local=$MAVEN_USER_HOME
  artifacts:
    reports:
      junit: target/test-results/**/TEST-*.xml
    paths:
      - target/test-results
      - target/jacoco
    expire_in: 1 day

frontend-test:
  stage: test
  script:
    - ./mvnw -ntp com.github.eirslett:frontend-maven-plugin:npm -Dfrontend.npm.arguments='run test' -Dmaven.repo.local=$MAVEN_USER_HOME
  artifacts:
    reports:
      junit: target/test-results/TESTS-results-jest.xml
    paths:
      - target/test-results
      - target/jacoco
    expire_in: 1 day

maven-package:
  stage: package
  script:
    - ./mvnw -ntp verify -Pprod -DskipTests -Dmaven.repo.local=$MAVEN_USER_HOME
  artifacts:
    paths:
      - target/*.jar
      - target/classes
    expire_in: 1 day

# Uncomment the following line to use gitlabs container registry. You need to adapt the REGISTRY_URL in case you are not using gitlab.com
docker-push:
  stage: release
  variables:
    REGISTRY_URL: registry.gitlab.com
    IMAGE_TAG: $CI_REGISTRY_IMAGE:$CI_COMMIT_REF_SLUG-$CI_COMMIT_SHA
  dependencies:
    - maven-package
  script:
    - ./mvnw -ntp compile jib:build -Pprod -Djib.to.image=$IMAGE_TAG -Djib.to.auth.username=gitlab-ci-token  -Djib.to.auth.password=$CI_BUILD_TOKEN -Dmaven.repo.local=$MAVEN_USER_HOME

```


##### **Browsers and Operating System**

Mac OS, but not relevant

-   [X] Checking this box is mandatory (this is just to show you read everything)

<!-- Love JHipster? Please consider supporting our collective:
👉  https://opencollective.com/generator-jhipster/donate -->

--- a/generators/ci-cd/templates/.gitlab-ci.yml.ejs
+++ b/generators/ci-cd/templates/.gitlab-ci.yml.ejs
@@ -19,7 +19,7 @@
 <%_ if (insideDocker) { _%>
 image: jhipster/jhipster:v<%= jhipsterVersion %>
 
-    <%_ if (cacheProvider === 'redis' || prodDatabaseType === 'cassandra' || prodDatabaseType === 'couchbase' || prodDatabaseType === 'neo4j') { _%>
+    <%_ if (cacheProvider === 'redis' || ['cassandra', 'couchbase', 'neo4j'].includes(prodDatabaseType)) { _%>
 # DinD service is required for Testcontainers
 services:
     - docker:dind

--- a/generators/ci-cd/templates/.gitlab-ci.yml.ejs
+++ b/generators/ci-cd/templates/.gitlab-ci.yml.ejs
@@ -18,6 +18,18 @@
 -%>
 <%_ if (insideDocker) { _%>
 image: jhipster/jhipster:v<%= jhipsterVersion %>
+
+    <%_ if (cacheProvider === 'redis' || prodDatabaseType === 'cassandra' || prodDatabaseType === 'couchbase' || prodDatabaseType === 'neo4j') { _%>
+# DinD service is required for Testcontainers
+services:
+    - docker:dind
+
+variables:
+    # Instruct Testcontainers to use the daemon of DinD.
+    DOCKER_HOST: 'tcp://docker:2375'
+    # Improve performance with overlayfs.
+    DOCKER_DRIVER: overlay2
+    <%_ } _%>
 <%_ } _%>
 
 <%_ if (buildTool === 'gradle') { _%>

